From 7ff14e13d0172666961e522b5b045fdaab0f8498 Mon Sep 17 00:00:00 2001
From: Laurentiu Palcu <laurentiu.palcu@nxp.com>
Date: Wed, 13 Jun 2018 09:36:10 +0300
Subject: [PATCH 3985/5242] MLK-18581: drm: imx: dcss: allow
 DRM_FORMAT_MOD_LINEAR modifier

commit  5c605314a281298aa8dbf28d9fc157c1275c8493 from
https://source.codeaurora.org/external/imx/linux-imx.git

Weston will use this modifier when playing linear formats so we need to
handle it correctly.

Signed-off-by: Laurentiu Palcu <laurentiu.palcu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/imx/dcss/dcss-plane.c |    9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/imx/dcss/dcss-plane.c b/drivers/gpu/drm/imx/dcss/dcss-plane.c
index 8617327..0a9fafb 100644
--- a/drivers/gpu/drm/imx/dcss/dcss-plane.c
+++ b/drivers/gpu/drm/imx/dcss/dcss-plane.c
@@ -62,6 +62,7 @@
 	DRM_FORMAT_MOD_VSI_G1_TILED,
 	DRM_FORMAT_MOD_VSI_G2_TILED,
 	DRM_FORMAT_MOD_VSI_G2_TILED_COMPRESSED,
+	DRM_FORMAT_MOD_LINEAR,
 	DRM_FORMAT_MOD_INVALID,
 };
 
@@ -150,7 +151,8 @@ static bool dcss_plane_format_mod_supported(struct drm_plane *plane,
 		case DRM_FORMAT_NV12:
 		case DRM_FORMAT_NV21:
 		case DRM_FORMAT_P010:
-			return modifier == DRM_FORMAT_MOD_VSI_G1_TILED ||
+			return modifier == DRM_FORMAT_MOD_LINEAR ||
+			       modifier == DRM_FORMAT_MOD_VSI_G1_TILED ||
 			       modifier == DRM_FORMAT_MOD_VSI_G2_TILED ||
 			       modifier == DRM_FORMAT_MOD_VSI_G2_TILED_COMPRESSED;
 		default:
@@ -352,6 +354,7 @@ static void dcss_plane_atomic_set_base(struct dcss_plane *dcss_plane)
 		break;
 	case DRM_PLANE_TYPE_OVERLAY:
 		if (!modifiers_present ||
+		    (modifiers_present && fb->modifier == DRM_FORMAT_MOD_LINEAR) ||
 		    (pix_format != DRM_FORMAT_NV12 &&
 		     pix_format != DRM_FORMAT_NV21 &&
 		     pix_format != DRM_FORMAT_P010)) {
@@ -485,6 +488,10 @@ static void dcss_plane_atomic_update(struct drm_plane *plane,
 	adj_w = src.x2 - src.x1;
 	adj_h = src.y2 - src.y1;
 
+	if (plane->type == DRM_PLANE_TYPE_OVERLAY &&
+	    modifiers_present && fb->modifier == DRM_FORMAT_MOD_LINEAR)
+		modifiers_present = false;
+
 	dcss_dpr_format_set(dcss_plane->dcss, dcss_plane->ch_num, pixel_format,
 				modifiers_present);
 	if (!modifiers_present)
-- 
1.7.9.5

